File: python/Add_Two_Numbers/Add_Two_Numbers.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        carry = 0
        # dummy head
        sum_val = l3 = ListNode(0)  #引用ListNode类定义一个链表节点并初始化值为0
        while l1 or l2:           #l1或l2不为空就持续执行
            val = carry             
            if l1:              #l1不为空，就把l1的某个节点值相加val再复制给val
                val += l1.val
                l1 = l1.next
            if l2:              #l2不为空，就把l2的某个节点值相加val再复制给val
                val += l2.val
                l2 = l2.next
            l3.next = ListNode(val % 10)
            l3 = l3.next
            carry = val / 10
        if carry > 0:
            l3.next = ListNode(carry)
        return sum_val.next

File: cpp/Add_Two_Numbers/Add_Two_Numbers.cpp

/// Source : https://leetcode.com/problems/add-two-numbers/description/
/// Author : bryce
/// Time   : 2019-10-13


/*
思路：

1.从最低位至最高位，逐位相加，如果和大于等于10，则保留个位数字，同时向前一位进1.
2.如果最高位有进位，则需在最前面补1.

做有关链表的题目，有个常用技巧：添加一个虚拟头结点：ListNode *head = new ListNode(0);，可以简化边界情况的判断。

时间复杂度：由于总共扫描一遍，所以时间复杂度是 O(n).
空间复杂度：O(n)

*/
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *sum = new ListNode(0);  //新建头结点，value为0，简化边界情况
        ListNode *l3=sum;
        int carry = 0, val1 = 0, val2 = 0, sum_val = 0;    //carry用于存储进位，sum_val存储相加                   
        while(l1!=NULL||l2!=NULL)            //三种情况：都不为空链表 1、等长，2、不等长； 有一个为空链表；
        {
            val1 = (l1!=NULL)?l1->val:0;
            val2 = (l2!=NULL)?l2->val:0;
            sum_val = (val1 + val2 + carry) % 10;
            carry = (val1 + val2 + carry)/10; //进位为0或大于1
            
            l3->next = new ListNode (sum_val);
            l3 = l3->next;
            
            if(l1!=NULL) l1=l1->next;
            if(l2!=NULL) l2=l2->next;
        }
        
        if(carry >= 1)
        {
            l3->next = new ListNode(carry);   //不要忽略最后一位的进位
        }
        
        return sum->next;                  //返回除头节点以外数
        
    }
};
